src/librustc_codegen_llvm/coverageinfo/mapgen.rs

```diff
@@ -0,0 +1,197 @@
+                            region_loc.start_line,
+                            region_loc.start_col,
+                            region_loc.end_line,
+                            region_loc.end_col,
+                        );
+                    }
+                    CoverageKind::CounterExpression(lhs, op, rhs) => {
+                        debug!(
+                            "  CounterExpression {} = {} {:?} {}, region_loc: {:?}",
+                            index, lhs, op, rhs, region_loc,
+                        );
+                        mapping_regions.push_from(
+                            *index,
+                            file_id,
+                            region_loc.start_line,
+                            region_loc.start_col,
+                            region_loc.end_line,
+                            region_loc.end_col,
+                        );
+                        expressions.push_from(op, lhs, rhs);
+                    }
+                }
+            }
+            for unreachable in function_coverage_regions.unreachable_regions() {
+                debug!(
+                    "  Unreachable code region: {}..{}",
+                    unreachable.start_byte_pos, unreachable.end_byte_pos
+                );
+                bug!("Unreachable region not expected and not yet handled!")
+                // FIXME(richkadel): implement and call
+                //   mapping_regions.push_from(...) for unreachable regions
+            }
+
+            let virtual_file_mapping = sorted_files_for_mapping.values().cloned().collect();
+
+            // encode and append the current function's coverage mapping data
+            let old_len = coverage_mappings_buffer.len();
+            coverageinfo::write_mapping_to_buffer(
+                virtual_file_mapping,
+                expressions,
+                mapping_regions,
+                coverage_mappings_buffer,
+            );
+
+            // add a new function record
+            let name_ref = coverageinfo::compute_hash(&mangled_function_name);
+            let name_ref_val = cx.const_u64(name_ref);
+            let mapping_data_size_val =
+                cx.const_u32((coverage_mappings_buffer.len() - old_len) as u32);
+            let func_hash_val = cx.const_u64(coverageinfo.hash);
+            function_records.push(cx.const_struct(
+                &[name_ref_val, mapping_data_size_val, func_hash_val],
+                /*packed=*/ true,
+            ));
+        }
+    });
+
+    // encode all filenames covered in this module, ordered by `file_id`
+    let filenames_buffer = llvm::build_byte_buffer(|filenames_buffer| {
+        coverageinfo::write_filenames_section_to_buffer(filenames, filenames_buffer);
+    });
+
+    let mut coverage_size = coverage_mappings_buffer.len();
+    if coverage_size == 0 {
+        return;
+    }
+
+    // concatenate the ended filenames and encoded coverage mappings, and add additional zero bytes
```

Comment:
is "ended" a typo, or what does it mean in this context?

Reply:
yes, thanks!  ended -> encoded